### PR TITLE
Show assist actions also in Outline popup.

### PR DESCRIPTION
**Pull request by scheglov**
image

### CODE REVIEW

**Comment by devoncarew** [comment]
Awesome!

I think we can omit items from the menu instead of showing them as disabled (@InMatrix to confirm).

If we're going to show icons there, we should probably use more std 16x16 ones. I'm busy fiddling w/ the icons for the toolbar, so having 16x16 icons for this PR isn't necessary.

**Comment by scheglov** [comment]
PTAL
image

File: src/io/flutter/preview/PreviewView.java [hunk 1]
```diff
@@ -285,6 +311,40 @@ public void propertyChange(PropertyChangeEvent evt) {
     contentManager.setSelectedContent(content);
   }
 
+  private ActionGroup createTreePopupActions() {
+    // The corresponding tree item has just been selected.
+    // Wait short time for receiving assists from the server.
+    for (int i = 0; i < 20 && actionToChangeMap.isEmpty(); i++) {
```
Comment (devoncarew) [inline src/io/flutter/preview/PreviewView.java]: This seems practical, but I wonder if there are any other ways of doing this. Perhaps an IntelliJ utility API?

@jwren, who I believe has used an API to see it there are any event interrupts?

File: src/io/flutter/preview/PreviewView.java [hunk 2]
```diff
@@ -205,6 +222,15 @@ public void initToolWindow(@NotNull ToolWindow toolWindow) {
     tree.setCellRenderer(new OutlineTreeCellRenderer());
     tree.expandAll();
 
+    tree.addMouseListener(new PopupHandler() {
```
Comment (devoncarew) [inline src/io/flutter/preview/PreviewView.java]: Can we scope this to just items we know might have contributions? Nodes that represent widgets?

File: src/io/flutter/preview/PreviewView.java [hunk 3]
```diff
@@ -205,6 +222,15 @@ public void initToolWindow(@NotNull ToolWindow toolWindow) {
...
+      public void invokePopup(Component comp, int x, int y) {
+        final ActionManager actionManager = ActionManager.getInstance();
+        final ActionGroup group = createTreePopupActions();
+        final ActionPopupMenu popupMenu = actionManager.createActionPopupMenu(ActionPlaces.UNKNOWN, group);
+        popupMenu.getComponent().show(comp, x, y);
```
Comment (devoncarew) [inline src/io/flutter/preview/PreviewView.java]: If the group is empty, does anything show? Should we just not call toe show() method in those cases?

File: src/io/flutter/preview/PreviewView.java (hunk 1 above)
Comment (scheglov) [reply src/io/flutter/preview/PreviewView.java]: Unless there is anything asynchronous for popup menus (and I have not seen this anywhere yet), we have to wait with some timeout. There are different ways to implement this waiting, this is just one of them.

Comment (devoncarew) [reply src/io/flutter/preview/PreviewView.java]: Makes sense, I just wasn't sure whether there was some prior art in the IntelliJ APIs for helping us here. lgtm as is

File: src/io/flutter/preview/PreviewView.java (hunk 2 above)
Comment (scheglov) [reply src/io/flutter/preview/PreviewView.java]: Done.

File: src/io/flutter/preview/PreviewView.java (hunk 3 above)
Comment (scheglov) [reply src/io/flutter/preview/PreviewView.java]: Yes, we did show.
I'm fixing this.

**Comment by InMatrix** [comment]
This menu looks great!

**Pull request merged**